lib/ansible/module_utils/network/common/facts/facts.py

```diff
@@ -0,0 +1,130 @@
+#
+# -*- coding: utf-8 -*-
+# Copyright 2019 Red Hat
+# GNU General Public License v3.0+
+# (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+"""
+The junos facts base class
+this contains methods common to all facts subsets
+"""
+from ansible.module_utils.network.common.network import get_resource_connection
+from ansible.module_utils.six import iteritems
+
+
+class FactsBase(object):
+    """
+    The facts base class
+    """
+    def __init__(self, module):
+        self._module = module
+        self._warnings = []
+        self._gather_subset = module.params.get('gather_subset')
+        self._gather_network_resources = module.params.get('gather_network_resources')
+        self._connection = get_resource_connection(module)
+
+        self.ansible_facts = {'ansible_network_resources': {}}
+        self.ansible_facts['ansible_gather_network_resources'] = list()
+        self.ansible_facts['ansible_net_gather_subset'] = list()
+
+        if not self._gather_subset:
+            self._gather_subset = ['!config']
+        if not self._gather_network_resources:
+            self._gather_network_resources = ['all']
+
+    def gen_runable(self, subsets, valid_subsets):
+        """ Generate the runable subset
+
+        :param module: The module instance
+        :param subsets: The provided subsets
+        :param valid_subsets: The valid subsets
+        :rtype: list
+        :returns: The runable subsets
+        """
+        runable_subsets = set()
+        exclude_subsets = set()
+        minimal_gather_subset = frozenset(['default'])
+
+        for subset in subsets:
+            if subset == 'all':
+                runable_subsets.update(valid_subsets)
+                continue
+            if subset == 'min' and minimal_gather_subset:
+                runable_subsets.update(minimal_gather_subset)
+                continue
+            if subset.startswith('!'):
+                subset = subset[1:]
+                if subset == 'min':
+                    exclude_subsets.update(minimal_gather_subset)
+                    continue
+                if subset == 'all':
+                    exclude_subsets.update(
+                        valid_subsets - minimal_gather_subset)
+                    continue
+                exclude = True
+            else:
+                exclude = False
+
+            if subset not in valid_subsets:
+                self._module.fail_json(msg='Bad subset')
+
+            if exclude:
+                exclude_subsets.add(subset)
+            else:
+                runable_subsets.add(subset)
+
+        if not runable_subsets:
+            runable_subsets.update(valid_subsets)
+        runable_subsets.difference_update(exclude_subsets)
+
+        return runable_subsets
+
+    def get_network_resources_facts(self, netres_choices, facts_resource_obj_map, resource_facts_type=None, data=None):
+        """
+        :param netres_choices:
```

Comment:
can we rename it as:
```suggestion
        :param net_res_choices:
```